### PR TITLE
Refactors projectile dampeners and adds some new visuals

### LINKED ISSUE #77093: Peacekeeper borgs' fields get dragged away in space
## Round ID:
[210999](https://scrubby.melonmesa.com/round/210999)

## Testmerges:
N/A

## Reproduction:
Basically #76172 , but with the blue square fields that peacekeeper borgs can create - if they stand somewhere near an edge of the shuttle, part of the field gets dragged away
field

### LINKED ISSUE #83682: Hyperkinetic Dampening projector (peacekeeper cyborg) invisible after activation
## Reproduction:

dreamseeker_HzxFKPpSnq

### LINKED ISSUE #84793: Projectile Dampener effect does NOT go away
## Issue Summary

The projectile dampener once turned on, can be turned off simply from quick menu or MOD settings. However, its effect around the user does not go away. With the suit turned off, wearer fully naked, it still stays on. The actual effect might be gone, but the visuals stay there.

## Round ID:

NA

## Testmerges:

NA, reproducable in Local master at latest commit: 59353dcc55f153f647106653bf05a86da0d19bff as of now.

## Reproduction:

1.  Get a modsuit with projectile dampener on(like HOS Safeguard)
2. Activate Projectile dampener from either quick menu or MOD settings.
3. Turn it off from either.
4. Dampener visuals remain constant.
5. Take off mod and other clothing, field continues to go with you.
6. Visible by everyone.

https://github.com/tgstation/tgstation/assets/68121607/6319e378-87cc-47cb-b622-159dec9683fe

Sorry for the abhorrent quality near the end, I had to compress my 40MB recording down to 5MB for github to allow uploading!

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
this pr completely rewrites projectile dampener logic, to make it look alot smoother and for performance's sake. Here's a before vs after of the field visuals movement.
 
before:

https://github.com/user-attachments/assets/887bfe85-2433-4a3b-8ded-a4013242d7dd

after:

https://github.com/user-attachments/assets/05388bcc-96c5-4f4a-b186-723800042e32

ive also added some new visual effects when a bullet enters the field, where itll sort of glitch out

https://github.com/user-attachments/assets/daef8053-4697-41af-8634-c7a88ff3d9fd

## Why It's Good For The Game
this improves projectile dampener performance by a significant much. here's a cost comparison after 6000 steps with each activated (left-side is before, right is after)
image

also i noticed that projectile dampeners were looping over range() on fast processing. which defeats the purpose of a proximity monitor to begin with. this pr removes the need for that
closes #84793

## Changelog
:cl:
refactor: projectile dampeners have been refactored. please report any bugs 
/:cl:

### CODE REVIEW

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+	var/static/image/edgeturf_south = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_south")
+	var/static/image/edgeturf_north = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_north")
+	var/static/image/edgeturf_west = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_west")
+	var/static/image/edgeturf_east = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_east")
+	var/static/image/northwest_corner = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_northwest")
+	var/static/image/southwest_corner = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_southwest")
+	var/static/image/northeast_corner = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_northeast")
+	var/static/image/southeast_corner = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_southeast")
+	var/static/image/generic_edge = image('icons/effects/fields.dmi', icon_state = "projectile_dampen_generic")
```
Comment (SmArtKar) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm:19]: this is really ugly, you recon you can turn this into an assoc list with bitflags?

**Review by SmArtKar — CHANGES_REQUESTED** [review]
Looks sane to me? But please doc projectile_dampener.dm, there's no comments and its quite complex

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	my_movable.Move(source.loc, get_dir(my_movable.loc, source.loc), source.glide_size)
+
+/datum/proximity_monitor/advanced/projectile_dampener/on_z_change(datum/source)
+	recalculate_field(full_recalc = TRUE)
+
+/datum/proximity_monitor/advanced/projectile_dampener/field_edge_crossed(atom/movable/movable, turf/location, turf/old_location)
+	. = ..()
+	if(!isprojectile(movable))
+		return
+	determine_wobble(location)
```
[the remark this answers is not in the document]
Comment (Fikou) [reply code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm:50]: the projectile is not holographic

**Comment by Sealed101** [comment]
some issues
- #77093
- #83682

**Comment by Ben10Omintrix** [comment]
> some issues
> 
> * [Peacekeeper borgs' fields get dragged away in space #77093](https://github.com/tgstation/tgstation/issues/77093)
> * [Hyperkinetic Dampening projector (peacekeeper cyborg) invisible after activation #83682](https://github.com/tgstation/tgstation/issues/83682)

just tested these. seems they arent an issue anymore

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	if(!iscyborg(movable) || !HAS_TRAIT_FROM(movable, TRAIT_GOT_DAMPENED, REF(src)))
+		ADD_TRAIT(movable, TRAIT_GOT_DAMPENED, REF(src))
+
+/datum/proximity_monitor/advanced/projectile_dampener/peaceborg/field_turf_uncrossed(atom/movable/movable, turf/old_location, turf/new_location)
+	if(!iscyborg(movable) || get_dist(new_location, host) <= current_range)
+		return
+	REMOVE_TRAIT(movable, TRAIT_GOT_DAMPENED, REF(src))
+
+/datum/proximity_monitor/advanced/projectile_dampener/peaceborg/setup_field_turf(turf/target)
+	for(var/atom/interesting_atom in target)
```
Comment (Ghommie) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm:198]:
```suggestion
	for(var/atom/interesting_atom as anything in target)
```

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+		if(iscyborg(interesting_atom))
+			ADD_TRAIT(interesting_atom, TRAIT_GOT_DAMPENED, REF(src))
+		if(isprojectile(interesting_atom))
+			catch_bullet_effect(interesting_atom)
+
+/datum/proximity_monitor/advanced/projectile_dampener/peaceborg/cleanup_field_turf(turf/target)
+	for(var/atom/interesting_atom in target)
```
Comment (Ghommie) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener.dm:205]:
```suggestion
	for(var/atom/interesting_atom as anything in target)
```

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		bullet.damage *= projectile_damage_multiplier
+	if(projectile_knockdown_multiplier)
+		bullet.knockdown *= projectile_knockdown_multiplier
+	if(projectile_stamina_multiplier)
+		bullet.stamina *= projectile_stamina_multiplier
+	if(projectile_stun_multiplier)
+		bullet.stun *= projectile_stun_multiplier
+
+/datum/dampener_projectile_effects/proc/remove_effects(obj/projectile/bullet)
+	bullet.speed = bullet::speed
```
Comment (Ghommie) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm:27]: These are going to cause some quirks with projectiles that assume different properties and values on runtime.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 6]
```diff
@@ -1053,3 +1058,17 @@
...
 	update_stat()
+
+/mob/living/silicon/robot/proc/on_dampen()
+	SIGNAL_HANDLER
+	eject_riders()
+
+/mob/living/silicon/robot/proc/eject_riders()
+	if(!length(buckled_mobs))
+		return
+	for(var/mob/living/buckled_mob in buckled_mobs)
```
Comment (Ghommie) [inline code/modules/mob/living/silicon/robot/robot.dm:1069]: As anything...

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm [hunk 7]
```diff
@@ -0,0 +1,38 @@
+///datum that holds the effects we have on bullet catching
+/datum/dampener_projectile_effects
+	///new projectiles speeds
+	var/projectile_speed_multiplier = 2.5
```
Comment (SmArtKar) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm:4]: Should be inverted to
```suggestion
	var/projectile_speed_multiplier = 0.4
```

now that #87740 is merged

File: code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+/datum/dampener_projectile_effects/proc/remove_effects(obj/projectile/bullet)
+	bullet.speed = bullet::speed
+	bullet.damage = bullet::damage
+	bullet.knockdown = bullet::knockdown
+	bullet.stamina = bullet::stamina
+	bullet.stun = bullet::stun
+
+/datum/dampener_projectile_effects/peacekeeper
+	projectile_speed_multiplier = 1.5
```
Comment (SmArtKar) [inline code/datums/proximity_monitor/fields/projectile_dampener/projectile_dampener_effects.dm:34]:
```suggestion
	projectile_speed_multiplier = 0.66
```

**Review by SmArtKar — CHANGES_REQUESTED** [review]

**Pull request merged**